syntax = "proto3";

package confidence.flags.resolver.v1;

option java_package = "com.spotify.confidence.flags.resolver.v1";
option java_multiple_files = true;
option java_outer_classname = "ApiProto";

import "google/protobuf/struct.proto";
import "types.proto";
import "confidence/flags/resolver/v1/types.proto";



message ResolveSimpleRequest {
    string client_secret = 1;
    google.protobuf.Struct evaluation_context = 2;
    string name = 3;
}

message ResolveFlagsRequest {
  // If non-empty, the specific flags are resolved, otherwise all flags
  // available to the client will be resolved.
  repeated string flags = 1;

  // An object that contains data used in the flag resolve. For example,
  // the targeting key e.g. the id of the randomization unit, other attributes
  // like country or version that are used for targeting.
  google.protobuf.Struct evaluation_context = 2;

  // Credentials for the client. It is used to identify the client and find
  // the flags that are available to it.
  string client_secret = 3;

  // Determines whether the flags should be applied directly as part of the
  // resolve, or delayed until `ApplyFlag` is called. A flag is typically
  // applied when it is used, if this occurs much later than the resolve, then
  // `apply` should likely be set to false.
  bool apply = 4;

  // Information about the SDK used to initiate the request.
  confidence.flags.resolver.v1.Sdk sdk = 5;
}

message ResolveFlagsResponse {
  // The list of all flags that could be resolved. Note: if any flag was
  // archived it will not be included in this list.
  repeated ResolvedFlag resolved_flags = 1;

  // An opaque token that is used when `apply` is set to false in `ResolveFlags`.
  // When `apply` is set to false, the token must be passed to `ApplyFlags`.
  bytes resolve_token = 2;

  // Unique identifier for this particular resolve request.
  string resolve_id = 3;
}


message ResolvedFlag {
  // The id of the flag that as resolved.
  string flag = 1;

  // The id of the resolved variant has the format `flags/abc/variants/xyz`.
  string variant = 2;

  // The value corresponding to the variant. It will always be a json object,
  // for example `{ "color": "red", "size": 12 }`.
  google.protobuf.Struct value = 3;

  // The schema of the value that was returned. For example:
  // ```
  // {
  //    "schema": {
  //      "color": { "stringSchema": {} },
  //      "size": { "intSchema": {} }
  //    }
  // }
  // ```
  // types.v1.FlagSchema.StructFlagSchema flag_schema = 4;

  // The reason to why the flag could be resolved or not.
  rust_guest.ResolveReason reason = 5;
}

// Request for resolving flags with sticky (materialized) assignments
message ResolveWithStickyRequest {
  // The standard resolve request
  ResolveFlagsRequest resolve_request = 1;

  // if we should support sticky or completely skip the flag if they had sticky rules
  bool not_process_sticky = 4;

  // Context about the materialization required for the resolve
  repeated ReadResult materializations = 5;

  reserved 3;
}

// Response from resolving with sticky assignments
message ResolveWithStickyResponse {
  oneof resolve_result {
    Success success = 1;
    ReadOperationsRequest read_ops_request = 3;
  }
  
  // Successful resolution with materialization updates
  message Success {
    ResolveFlagsResponse response = 1;
    // New assignments that should be stored
    repeated VariantData materialization_updates = 3;
  }
}

message VariantReadOp {
  string unit = 1;
  string materialization = 2;
  string rule = 3;
}

message InclusionReadOp {
  string unit = 1;
  string materialization = 2;
}

message ReadOp {
  oneof op {
    VariantReadOp variant_read_op = 1;
    InclusionReadOp inclusion_read_op = 2;
  }
}

message ReadOperationsRequest {
  repeated ReadOp ops = 3;
}

message VariantData {
  string unit = 1;
  string materialization = 2;
  string rule = 3;
  string variant = 4;
}

message InclusionData {
  string unit = 1;
  string materialization = 2;
  bool is_included = 3;
}

message ReadResult {
  oneof result {
    VariantData variant_result = 1;
    InclusionData inclusion_result = 2;
  }
}

message ReadOperationsResult {
  repeated ReadResult results = 1;
}